jplag/src/main/java/de/jplag/clustering/algorithm/AgglomerativeClustering.java

```diff
@@ -0,0 +1,189 @@
+            for (Cluster otherCluster : clusters) {
+                similarities.add(new ClusterConnection(combined, otherCluster, similarityMatrix));
+            }
+            clusters.add(combined);
+        }
+
+        return clusters.stream().map(Cluster::getSubmissions).collect(Collectors.toList());
+    }
+
+    private float clusterSimilarity(List<Integer> leftCluster, List<Integer> rightCluster, RealMatrix similarityMatrix) {
+        float similarity = 0;
+        switch (options.similarity) {
+            case MIN:
+                similarity = Float.MAX_VALUE;
+                break;
+            case MAX:
+                similarity = -Float.MAX_VALUE;
+                break;
+            case AVERAGE:
+                similarity = 0;
+                break;
+        }
+
+        for (int leftIndex = 0; leftIndex < leftCluster.size(); leftIndex++) {
+            int leftSubmission = leftCluster.get(leftIndex);
+            for (int rightIndex = 0; rightIndex < rightCluster.size(); rightIndex++) {
+                float submissionSimilarity = (float) similarityMatrix.getEntry(leftSubmission, rightCluster.get(rightIndex));
+                switch (options.similarity) {
+                    case MIN:
+                        similarity = Math.min(similarity, submissionSimilarity);
+                        break;
+                    case MAX:
+                        similarity = Math.max(similarity, submissionSimilarity);
+                        break;
+                    case AVERAGE:
+                        similarity += submissionSimilarity;
+                }
+            }
+        }
+
+        if (options.similarity == InterClusterSimilarity.AVERAGE) {
+            similarity /= leftCluster.size() * rightCluster.size();
+        }
+        return similarity;
+    }
+
+    public static class ClusteringOptions {
+        public float minimalSimilarity = 0.1f;
+        public InterClusterSimilarity similarity = InterClusterSimilarity.AVERAGE;
+    }
+
+    public static enum InterClusterSimilarity {
+        MIN,
+        MAX,
+        AVERAGE;
+    }
+
+    private class ClusterConnection implements Comparable<ClusterConnection> {
+
+        private Cluster left;
+        private Cluster right;
+        private float similarity;
+
+        public ClusterConnection(Cluster left, Cluster right, RealMatrix similarity) {
+            this(left, right, clusterSimilarity(left.submissions, right.submissions, similarity));
+        }
+
+        public ClusterConnection(Cluster left, Cluster right, float similarity) {
+            this.left = left;
+            this.right = right;
+            this.similarity = similarity;
+        }
+
+        @Override
+        public int compareTo(ClusterConnection other) {
+            return (int) Math.signum(other.similarity - similarity);
+        }
+
+    }
+
+    private class Cluster {
```

Comment:
Because this is juggled around in a hash maps and -sets.
Lists usually implement hashing based on their members, which would impact performance.
Adding a comment.